dynamicio/base.py

```diff
@@ -0,0 +1,126 @@
+# pylint: disable=no-name-in-module disable=invalid-name
+
+"""BaseResource class for creating various resources types."""
+from abc import ABC, abstractmethod
+from typing import Optional, Type, TypeVar
+
+import pandas as pd
+import pandera as pa
+from pandera import SchemaModel
+from pydantic import BaseModel
+
+SchemaType = TypeVar("SchemaType", bound=pa.SchemaModel)  # TODO: utilise this
+
+
+class BaseResource(BaseModel, ABC):
+    """BaseClass for resource classes."""
+
+    pa_schema: Optional[pa.SchemaModel]
+    validate_default: bool = True
+    log_metrics_default: bool = True
+    allow_no_schema: bool = False
+
+    def read(
+        self,
+        validate: Optional[bool] = None,
+        log_metrics: Optional[bool] = None,
+        pa_schema: Optional[Type[SchemaModel]] = None,
+    ) -> pd.DataFrame:
```

Comment:
In `v5` should we have a discussion whether or not we should hardcode the implementation to specifically to Pandas dataframes? Would it be difficult to migrate to, e.g., Spark/Polars dataframes in the future? I'm not saying we should overengineer, but I'm wondering how hard this switch could be

Reply:
I think the switch would be quite easy, but with separate code following the exact same pattern. For example, polars and [patito](https://github.com/kolonialno/patito) would be the equivalent to pandas and pandera. Once this part is released/simplified even more we can maybe see whether generalising it would be an option.